pkg/loki/config_wrapper.go

```diff
@@ -96,85 +99,141 @@ func (c *ConfigWrapper) ApplyDynamicConfig() cfg.Source {
 	}
 }
 
-// applyIngesterRingConfig will use whatever config is setup for the ingester ring and use it everywhere else
-// we have a ring configured. The reason for centralizing on the ingester ring as this is been set in basically
-// all of our provided config files for all of time, usually set to `inmemory` for all the single binary Loki's
-// and is the most central ring config for Loki.
+// applyDynamicRingConfigs checks the current config and, depending on the given values, reuse ring configs accordingly.
 //
-// If the ingester ring has its interface names sets to a value equal to the default (["eth0", en0"]), it will try to append
-// the loopback interface at the end of it.
-func applyIngesterRingConfig(r *ConfigWrapper, defaults *ConfigWrapper) error {
-	if reflect.DeepEqual(r.Ingester.LifecyclerConfig.InfNames, defaults.Ingester.LifecyclerConfig.InfNames) {
-		appendLoopbackInterface(r)
+// 1. Gives preference to any explicit ring config set. For instance, if the user explicitly configures Distributor's ring,
+// that config will prevail. This rule is enforced by the fact that the config file and command line args are parsed
+// again after the dynamic config has been applied, so will take higher precedence.
+// 2. If no explicit ring config is set, use the common ring configured if provided.
+// 3. If no common ring was provided, use the memberlist config if provided.
+// 4. If no common ring or memberlist were provided, use the ingester's ring configuration.
+
+// When using the ingester or common ring config, the loopback interface will be appended to the end of
+// the list of default interface names
+func applyDynamicRingConfigs(r, defaults *ConfigWrapper) {
+	if reflect.DeepEqual(r.Common.Ring, defaults.Common.Ring) {
+		// common ring not set, so use memberlist for all rings if set
+		if len(r.MemberlistKV.JoinMembers) > 0 {
+			applyMemberlistConfig(r)
+		} else {
+			// neither common ring nor memberlist set, use ingester ring configuration for all rings
+			useIngesterRingConfig(r, defaults)
+		}
+	} else {
+		// common ring is provided, use that for all rings
+		useCommonRingConfig(r, defaults)
 	}
 
-	lc := r.Ingester.LifecyclerConfig
-	rc := r.Ingester.LifecyclerConfig.RingConfig
-	s := rc.KVStore.Store
-	sc := r.Ingester.LifecyclerConfig.RingConfig.KVStore.StoreConfig
+}
 
-	f, err := tokensFile(r, "ingester.tokens")
-	if err != nil {
-		return err
-	}
-	r.Ingester.LifecyclerConfig.TokensFilePath = f
+func useCommonRingConfig(r, defaults *ConfigWrapper) {
+	// if the default ingester config is detected, then it's safe to override with the common config
+	shouldOverrideIngester := reflect.DeepEqual(r.Ingester.LifecyclerConfig, defaults.Ingester.LifecyclerConfig)
```

Comment:
@owen-d per your comment on @DylanGuedes's original PR, would you like to see this `DeepEqual` check applied to all ring overrides? The reason why we may not need to do that is because after this `ApplyDynamicConfig` function runs, we the parse the config file and command line args again. So, if someone is not using the default, they must have provided configs in one of those 2 places, and thus they will get applied and overrides any dynamic configs we applied.

Reply:
That's a good point. I don't think it's necessary 👍

Reply:
Hey, since in such a scenario the configs would be applied again, maybe this whole `shouldOverrideIngester` isn't necessary?

Reply:
hmm, and just always override the ingester. that's a good point. I believe we have test coverage for that case so I'll try it out